### PR TITLE
feat(msk): Cluster L2 Construct

### PR DESCRIPTION
**Pull request by curtiseppel**
L2 Construct for a MSK Cluster. 

I wrote this for internal use and thought I'd share it. I tried to follow the [example resource](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/example-construct-library/lib/example-resource.ts) and [design guidelines](https://github.com/aws/aws-cdk/blob/master/DESIGN_GUIDELINES.md) as much as I could. Default properties were chosen either based on defaults when creating a cluster in the console or defaults set from CloudFormation.

Closes https://github.com/aws/aws-cdk/issues/9603

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by CurtisEppel** [comment]
@iliapolo I've updated this PR to support the latest Kafka version and to support MSK's new authentication method - SASL/SCRAM. I also added a bunch of tests and the build is now passing.

**Comment by iliapolo** [comment]
@CurtisEppel Thats awesome, thanks! Apologies the review is little lagging, its still on my radar though.

File: packages/@aws-cdk/aws-msk/lib/cluster-props.ts
Comment (CurtisEppel) [inline packages/@aws-cdk/aws-msk/lib/cluster-props.ts:172]: I'm not sure if an enum is a good fit for this. This is the second time I've had to update this enum as a new monitoring level has been released (https://aws.amazon.com/about-aws/whats-new/2020/11/amazon-msk-offers-consumer-lag-metrics-select-topic-level-metrics-free/). Maybe should replace it with a class like [KafkaVersion](https://github.com/aws/aws-cdk/blob/ba75c8b45a517274cbc1874efce1ce1dd2c926f1/packages/%40aws-cdk/aws-msk/lib/cluster-version.ts) so that users can escape the enum with a custom value.

**Review by iliapolo — DISMISSED** [review]
@CurtisEppel  I had several comments asking to flatten out configuration properties, in contrast to the CloudFormation structure. I just want to clarify the reasoning here.

Excessive nesting makes it hard to read and reason about, I know it allows for future proofing, but we normally don't do that. From prior experience, it's usually better to provide the best experience possible and employ deprecations in the future if necessary, than to sacrifice the ergonomics of the current API for some future unknown capabilities. 

Thanks for this PR!

File: packages/@aws-cdk/aws-msk/lib/cluster.ts
Comment (CurtisEppel) [reply packages/@aws-cdk/aws-msk/lib/cluster.ts:45]: This was taken directly from the example - https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/example-construct-library/lib/example-resource.ts#L165

File: packages/@aws-cdk/aws-msk/lib/cluster.ts
Comment (CurtisEppel) [reply packages/@aws-cdk/aws-msk/lib/cluster.ts:160]: > Amazon MSK always encrypts your data at rest.
https://docs.aws.amazon.com/msk/latest/developerguide/msk-encryption.html#msk-encryption-at-rest

File: packages/@aws-cdk/aws-msk/lib/cluster.ts
Comment (CurtisEppel) [reply packages/@aws-cdk/aws-msk/lib/cluster.ts:215]: An alias would be nice, how about including the cluster name in the alias?

Side note the alias is actually not that useful as the MSK API doesn't allow you to associate secrets with the cluster that were created using the alias, even though Secrets Manager allows them to be created that way. I've created a support ticket and it has been verified as a known bug.

File: packages/@aws-cdk/aws-msk/lib/cluster.ts
Comment (CurtisEppel) [reply packages/@aws-cdk/aws-msk/lib/cluster.ts:231]: This policy gives users permission to use the key for the actions listed only when Secrets Manager makes the request on the user's behalf and doesn't allow any user to use the key directly. This means the resources can be `*` and the resources will instead be scoped by the users role if they have permission to read/write secrets.

File: packages/@aws-cdk/aws-msk/lib/cluster.ts
Comment (CurtisEppel) [reply packages/@aws-cdk/aws-msk/lib/cluster.ts:317]: Good to know about lazy instances. Thanks :)

File: packages/@aws-cdk/aws-msk/lib/cluster.ts
Comment (CurtisEppel) [reply packages/@aws-cdk/aws-msk/lib/cluster.ts:218]: https://docs.aws.amazon.com/kms/latest/developerguide/services-secrets-manager.html#asm-policies

File: packages/@aws-cdk/aws-msk/lib/cluster.ts
Comment (CurtisEppel) [reply packages/@aws-cdk/aws-msk/lib/cluster.ts:1]: Can you clarify what should be moved to `cluster.ts`?

Comment (iliapolo) [reply packages/@aws-cdk/aws-msk/lib/cluster.ts:1]: Whoops. I meant to put this comment on the `cluster-props.ts` file. The properties of the cluster should be in the same file as the cluster.

File: packages/@aws-cdk/aws-msk/lib/cluster.ts
Comment (CurtisEppel) [reply packages/@aws-cdk/aws-msk/lib/cluster.ts:455]: Nope you're right - it's simpler to just check if PLAINTEXT and clientAuthenticationConfiguration is provided

**Comment by bracki** [comment]
Hey @iliapolo, @CurtisEppel, I have worked on MSK/Kafka event sources for Lambda here: https://github.com/aws/aws-cdk/issues/12099

Maybe we could have quick chat on the CDK slack or similar because I'm looking for feedback and because it would make sense if the EventSource can work with your implementation.
I could also provide code on how to setting up user/password-authentication with a AwsCustomResource for this PR. I'm "@bracki" on Slack. Thanks!

**Comment by iliapolo** [comment]
@bracki Lets just have the discussion here, as i suspect it will be highly asynchronous and take a few days, this is a better place to keep track of it.

You are welcome to start it :)

**Comment by bracki** [comment]
OK, my PR is ready and green. But it probably makes sense to wait for this PR to get ready. Currently I depend on `clusterArn:  string` but it makes more sense to take a Kafka cluster as an argument.

What does it take to get this merged?

**Comment by iliapolo** [comment]
@bracki Yes I agree it makes sense to wait and use the L2 instead. This PR is under active review and I think should be merged fairly soon.

**Comment by iliapolo** [comment]
@CurtisEppel Let me know when its ready for a second round? You can just `re-request` a review.

**Comment by CurtisEppel** [comment]
> @CurtisEppel Let me know when its ready for a second round? You can just `re-request` a review.



> Hey @iliapolo, @CurtisEppel, I have worked on MSK/Kafka event sources for Lambda here: #12099
> 
> Maybe we could have quick chat on the CDK slack or similar because I'm looking for feedback and because it would make sense if the EventSource can work with your implementation.
> I'm "@bracki" on Slack. Thanks!

Nice, I also had an implementation of this but was holding off to add it to a separate PR once this one was through. I've been busy but will try and get this PR finished this week so that we can integrate your event source with this L2 construct.

File: packages/@aws-cdk/aws-msk/lib/cluster.ts
Comment (CurtisEppel) [reply packages/@aws-cdk/aws-msk/lib/cluster.ts:543]: I've included the cluster name

File: packages/@aws-cdk/aws-msk/lib/cluster-props.ts
Comment (CurtisEppel) [reply packages/@aws-cdk/aws-msk/lib/cluster-props.ts]: Mainly following how CloudFormation grouped their properties. I'll see what flattening would look like.

> Are there any other nodes expect for brokers?
Not that I know of

File: packages/@aws-cdk/aws-msk/test/cluster.test.ts
Comment (iliapolo) [inline packages/@aws-cdk/aws-msk/test/cluster.test.ts:21]: All tests under this `describe` can be rolled into one since they just assert on different parts of the template, without actually configuring anything. Just a single test with a big assertion on all the properties would be better.

File: packages/@aws-cdk/aws-msk/lib/cluster.ts
Comment (iliapolo) [inline packages/@aws-cdk/aws-msk/lib/cluster.ts:620]: Thats actually not what I meant. This would create a token each time it is called - and at synth time it would still create two constructs with the same id. 

Make sure you add a test for this.

What I meant is this:

```ts
  private _zookeeperConnectionString(responseField: string): string {
    if (!this._clusterDescription) {
      this._clusterDescription = new cr.AwsCustomResource(this, 'ZookeeperConnect', {
        onUpdate: {
          service: 'Kafka',
          action: 'describeCluster',
          parameters: {
            ClusterArn: this.clusterArn,
          },
          physicalResourceId: cr.PhysicalResourceId.of(
            'ZooKeeperConnectionString',
          ),
        },
        policy: cr.AwsCustomResourcePolicy.fromSdkCalls({
          resources: [this.clusterArn],
        }),
      });
    }
    return this._clusterDescription.getResponseField(`ClusterInfo.${responseField}`);
  }

```

I think it worth adding a `CfnOutput` for these attributes in the integ test, this would make sure the cr is invoked correctly.

**Review by iliapolo — DISMISSED** [review]
@CurtisEppel Notice there are also some `README` violations in the build. You need to run `yarn pkglint` to fix.

**Comment by iliapolo** [comment]
dont close

**Comment by iliapolo** [comment]
@CurtisEppel Just making sure - are you still looking at this PR?

**Comment by CurtisEppel** [comment]
> @CurtisEppel Just making sure - are you still looking at this PR?

Sorry @iliapolo, I've been busy at work with non-kafka things. I'll try and finish this off on this weekend.

File: packages/@aws-cdk/aws-msk/test/integ.cluster.ts
Comment (CurtisEppel) [reply packages/@aws-cdk/aws-msk/test/integ.cluster.ts]: I think I left this in from when i was testing the IConnectable locally. I've removed it.

**Comment by iliapolo** [comment]
@CurtisEppel When you're ready for another round just re-request a review 👍

**Comment by CurtisEppel** [comment]
> @CurtisEppel When you're ready for another round just re-request a review 👍

I have requested a review...

The latest build failed due to a Jest timeout after I merged the latest master. Is this something that needs to be retried or did I actually break something?

File: packages/@aws-cdk/aws-msk/lib/cluster.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-msk/lib/cluster.ts:382]: Actually, we can simply do:

```ts
public readonly clusterName = cdk.Fn.select(1, cdk.Fn.split('/', clusterArn));
```

To handle both tokens and concrete values.

File: packages/@aws-cdk/aws-msk/lib/cluster.ts
Comment (iliapolo) [inline packages/@aws-cdk/aws-msk/lib/cluster.ts:426]: Just throw instead of the annotation. Error annotations are normally used for validating lookup properties, i.e ones that can change outside of the synthesis process, this cannot. 

There are a couple of these around here, make sure you replace all of them.

File: packages/@aws-cdk/aws-msk/lib/cluster.ts
Comment (iliapolo) [inline packages/@aws-cdk/aws-msk/lib/cluster.ts:736]: Maybe add the cluster name? otherwise this will conflict when the same user is added to two different clusters.

**Review by iliapolo — CHANGES_REQUESTED** [review]

**Review by iliapolo — APPROVED** [review]
@CurtisEppel 

Awesome, thanks for all the hard work and sticking with this PR for so long :)

**Comment by iliapolo** [comment]
@CurtisEppel Looks like there are some build failures after rebasing from master. Have a look?

**Comment by CurtisEppel** [comment]
> @CurtisEppel
> 
> Awesome, thanks for all the hard work and sticking with this PR for so long :)

Thanks for all of the useful feedback 🎉 

@iliapolo  The snapshot test was failing due to lambda assets changing in the custom resource package. I've updated it now.

**Review by iliapolo — APPROVED** [review]

**Pull request merged**
